core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -932,7 +932,10 @@ public ClusterState execute(ClusterState currentState) {
                     if (modified) {
                         stateBuilder.nodes(nodesBuilder);
                     }
-                    return stateBuilder.build();
+                    currentState = stateBuilder.build();
```

Comment:
I wonder if we should shortcut if modified == false . I'm thinking about a join storm upon master failure which will lead to a series of "received a join request for an existing node " where we don't really need to do anything.